src/message.rs

```diff
@@ -166,6 +166,12 @@ where
     }
 }
 
+pub trait MessageDescriptor: Message + 'static {
+    fn message_name(&self) -> &'static str;
+    fn package_name(&self) -> &'static str;
+    fn type_url(&self) -> &'static str;
```

Comment:
We may not need `type_url` to be included in the trait. Let's take a look on how C++ implements `Any::PackFrom`:

```cpp
bool AnyMetadata::PackFrom(Arena* arena,
                           const Message& message,
                           StringPiece type_url_prefix) {
  type_url_->Set(
      &::PROTOBUF_NAMESPACE_ID::internal::GetEmptyString(),
      GetTypeUrl(message.GetDescriptor()->full_name(), type_url_prefix),
      arena);
  return message.SerializeToString(
      value_->Mutable(ArenaStringPtr::EmptyDefault{}, arena));
}
```

It uses an API called `Descriptor::full_name` which returns "<package>.<message>", and finally generates the type_url via `GetTypeUrl`, which is hard-coded in the library and can not be customized like a trait.

Reply:
The `type_url` was just added for convenience really. If this stops the PR from being merged then I'm all for removing it! A third-party library that implements pack and unpack like [prost-wkt](https://github.com/fdeantoni/prost-wkt) can then implement it instead (with the prefix configurable and whatnot).

Reply:
I really appreciate your work on prost-wkt and I'm actually not one of the prost maintainers, so I couldn't and didn't intend to stop anybody. I left this comment only because I feel a minimum public trait is better and will not lead to incompatibility issues in the future. Anyway, I'm really looking forward to this PR being merged, then we will not have to use a forked version of prost in prost-wkt any longer.